test/metabase/api/action_test.clj

```diff
@@ -305,3 +305,58 @@
+        (is (= 75
+               (categories-row-count)))
+        (is (= {:created-rows [{(format-field-name :id) 76, (format-field-name :name) "NEW_A"}
+                               {(format-field-name :id) 77, (format-field-name :name) "NEW_B"}]}
+               (mt/user-http-request :crowberto :post 200
+                                     (format "action/bulk/create/%d" (mt/id :categories))
+                                     [{(format-field-name :name) "NEW_A"}
+                                      {(format-field-name :name) "NEW_B"}])))
+        (is (= 77
+               (categories-row-count)))
+        (is (= {:success true}
+               (mt/user-http-request :crowberto :post 200
+                                     (format "action/bulk/delete/%d" (mt/id :categories))
+                                     [{"ID" 76}
+                                      {"ID" 77}])))
+        (is (= 75
+               (categories-row-count)))))))
+
+(deftest bulk-delete-failure-test
+  (testing "POST /api/action/bulk/delete/:table-id"
+    (mt/test-drivers (mt/normal-drivers-with-feature :actions)
+      (actions.test-util/with-actions-test-data-and-actions-enabled
+        (mt/test-drivers (mt/normal-drivers-with-feature :actions)
+          (actions.test-util/with-actions-test-data-and-actions-enabled
+            (testing "error in some of the rows"
+              (is (= 75
+                     (categories-row-count)))
+              (is (= {:created-rows [{(format-field-name :id) 76, (format-field-name :name) "NEW_A"}
+                                     {(format-field-name :id) 77, (format-field-name :name) "NEW_B"}]}
+                     (mt/user-http-request :crowberto :post 200
+                                           (format "action/bulk/create/%d" (mt/id :categories))
+                                           [{(format-field-name :name) "NEW_A"}
+                                            {(format-field-name :name) "NEW_B"}])))
+              (is (= 77
+                     (categories-row-count)))
+
+              (testing "Should report indices of bad rows"
+                (is (= {:errors
+                        [{:index 1,
+                          :error "Error filtering against :type/BigInteger Field: unable to parse String \"foo\" to a :type/BigInteger"}
+                         {:index 3,
+                          :error "Sorry, this would delete 0 rows, but you can only act on 1"}]}
+                       (mt/user-http-request :crowberto :post 400
+                                             (format "action/bulk/delete/%d" (mt/id :categories))
+                                             [{"ID" 76}
+                                              {"ID" "foo"}
+                                              {"ID" 77}
+                                              {"ID" 107}]))))
+              (is (= 77
+                     (categories-row-count))))))))))
```

Comment:
We **need** tests around the other validation you put in here. 

- Inconsistent keys passed in
- Keys not matching the PK columns for the Table

Otherwise it's entirely possible that the code that gets triggered on these errors gets broken somehow and we don't catch it because we have no coverage around it. That's exactly what happened with https://github.com/metabase/metabase/issues/24021 -- there wasn't any test coverage around `parse-error`/`parse-sql-error` so when I accidentally introduced a regression it went live without anyone knowing about it, and we only found out when someone manually tested stuff. Even a unit test that hit `parse-error` directly (outside of the API endpoints) would have been enough to catch the regression before we shipped it.

Reply:
On the flip side it would be possible for us to accidentally break or remove that validation and let people do truly busted or dangerous stuff. If those constraints are important to have we need to have tests making sure they're enforced